File: problems/Two Pointers/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        answer = []
        child = []
        nums.sort()

        for i in range(0, n-1):
            if(i != 0 and nums[i] == nums[i-1]):
                continue

            l = i + 1
            r = n - 1
            while l < r:
                if nums[i] + nums[l] + nums[r] == 0:
                    child.append(nums[i])
                    child.append(nums[l])
                    child.append(nums[r])
                    answer.append(list(child))
                    child.clear()
                    l += 1
                    r -= 1

                    while(l < r and nums[l] == nums[l-1]):
                        l += 1
                    while(l < r and nums[r] == nums[r+1]):
                        r -= 1

                elif nums[i] + nums[l] + nums[r] > 0:
                    r -= 1
                elif nums[i] + nums[l] + nums[r] < 0:
                    l += 1
        return answer

File: problems/Java/3sum.java
class Solution {
	public List<List<Integer>> threeSum(int[] nums) {
		int n = nums.length;
		List<List<Integer>> result = new ArrayList<>();
		Arrays.sort(nums);

		for (int i = 0; i < n; i++) {

			if (i != 0 && nums[i] == nums[i - 1]) {
				continue;
			}

			int l = i + 1;
			int r = n - 1;

			while (l < r) {
				int sum = nums[i] + nums[r] + nums[l];

				if (sum == 0) {
					result.add(Arrays.asList(nums[i], nums[r], nums[l]));

					l++;
					r--;

					while (l < r && nums[l] == nums[l - 1]) {
						l++;
					}
					while (l < r && nums[r] == nums[r + 1]) {
						r--;
					}

				} else if (sum > 0) {
					r--;
				} else if (sum < 0) {
					l++;
				}
			}
		}
		return result;
	}
}